### PR TITLE
New feature: Add `limit_render_to_text_bbox` to `Label` to limit text rendering to the text bounding box, improving `Label` alignments.

### PR DESCRIPTION
**Pull request by DexerBR**
This PR adds the `limit_render_to_text_bbox` property to the `Label` to limit the texture rendering to the text bounding box, allowing better control over positioning, and enabling more precise alignment of the Label in relation to other UI elements.

Currently there is no way adjust the label alignment properly, leading to imprecise results that are extremely difficult for the user to work around.

At the moment, support has been added only to the `PIL` text provider. But I believe that this feature should be expanded to other text providers in the future.

This PR is based on:
- https://github.com/kivy/kivy/pull/8511

### Demo:

https://github.com/kivy/kivy/assets/73297572/9cceb576-5881-4414-9b44-16ecae60568f

### Test code:

Note: Use "d" and "f" to change fonts.

```python
import os
os.environ['KIVY_TEXT'] = 'pil'

from kivy.lang import Builder
from kivy.app import App
from kivy.base import EventLoop


class MainApp(App):
    
    root_widget = Builder.load_string(r'''

FloatLayout
    BoxLayout:
        size_hint_y: None
        height: dp(100)
        orientation: "vertical"
        pos_hint: {"top": 1}
        TextInput:
            id: ti
            text: "a"
        ToggleButton:
            text: "limit_render_to_text_bbox"
            on_state:
                label_widget.limit_render_to_text_bbox = not self.state == "normal"

    FloatLayout:
        size_hint: 0.8, 0.5
        pos_hint: {"center_x": 0.5, "center_y": 0.5}
        canvas.before:
            Color:
                rgba: (0.5, 0.5, 0.5, 1)
            Rectangle:
                pos: self.pos
                size: self.size
            
            Color:
                rgba: (1, 0, 0, 1)
            Line:
                points: self.x, self.center_y, self.right, self.center_y
        
            
        Label:
            id: label_widget
            opacity: .5
            # text: (self.font_name.split('/')[-1], self.font_name)[0]
            text: ti.text
            font_size: sp(300)
            size_hint: None, None
            size: self.texture_size
            pos_hint: {"center_x": 0.5, "y": 0}
            limit_render_to_text_bbox: False

            canvas.before:
                Color:
                    rgba: (0, 0, 1, 1)
                Rectangle:
                    pos: self.pos
                    size: self.size
                                        
        BoxLayout:
            pos_hint: {"center_x": 0.5}
            size_hint_y: 0.25

            CheckBox:
                group: "pos"
                on_active:
                    label_widget.pos_hint = {"center_x": 0.5, "top": 1}
            CheckBox:
                group: "pos"
                on_active:
                    label_widget.pos_hint = {"center_x": 0.5, "center_y": 0.5}
            
            CheckBox:
                group: "pos"
                active: True
                active: True
                on_active:
                    label_widget.pos_hint = {"center_x": 0.5, "y": 0}

''')

    def build(self):
        EventLoop.window.bind(on_keyboard=self._hook_keyboard)
        return self.root_widget

    def _hook_keyboard(self, window, key, keycode, text, modifiers):
        # print('keyboard ays :', key, keycode, text, modifiers )

        if text == 'd':
            self.root.ids.label_widget.font_name = get_next_font(self.root.ids.label_widget.font_name, 'left')
        if text == 'f':
            self.root.ids.label_widget.font_name = get_next_font(self.root.ids.label_widget.font_name, 'right')


def get_font_list():
    import pathlib

    def get_fonts(fonts_path):
        '''Get a list of all the fonts available on this system.
        '''

        if not fonts_path: return []
        font_list = []
        for fdir in fonts_path:
            font_list += pathlib.Path(fdir).rglob('*.ttf')
        
        nflist = []
        for i, x in enumerate(font_list):
            if not str(x).split('/')[-1].startswith('._'):
                nflist.append(str(x))
        
        return sorted([*set(nflist)])
    
    from os.path import join, dirname, abspath
    from kivy.core.text import Label as CoreLabel

    return get_fonts(fonts_path = \
        [abspath(join(dirname(__file__), '..', 'data', 'fonts'))] +\
        CoreLabel.get_system_fonts_dir()
        )

font_list = get_font_list()

def get_next_font(pfont_name, l_r):
    try:
        idx = font_list.index(pfont_name)
    except ValueError:
        return font_list[0]

    try:
        font_name = str(font_list[idx +\
                                (-1 if l_r == 'left' else 1)])
    except IndexError:
        font_name = font_list[0]
    return font_name


if __name__ == '__main__':
    MainApp().run()
```

Maintainer merge checklist
* [x] Title is descriptive/clear for inclusion in release notes.
* [ ] Applied a `Component: xxx` label.
* [ ] Applied the `api-deprecation` or `api-break` label.
* [ ] Applied the `release-highlight` label to be highlighted in release notes.
* [ ] Added to the milestone version it was merged into.
* [ ] **Unittests** are included in PR.
* [ ] Properly documented, including `versionadded`, `versionchanged` as needed.

### CODE REVIEW

**Comment by Julian-O** [comment]
> FreeTypeFont.getsize() is deprecated since version 9.2.0 and [FreeTypeFont.getbbox()](https://pillow.readthedocs.io/en/stable/reference/ImageFont.html#PIL.ImageFont.FreeTypeFont.getbbox) should be used instead: https://pillow.readthedocs.io/en/stable/releasenotes/9.2.0.html

As discussed in #8427, the kivy-ios and python-for-android Pillow recipes depend on versions of 8.* versions of Pillow, where `getbbox` isn't available. I believe this change will make deployment to mobile devices impossible.

**Comment by DexerBR** [comment]
> > FreeTypeFont.getsize() is deprecated since version 9.2.0 and [FreeTypeFont.getbbox()](https://pillow.readthedocs.io/en/stable/reference/ImageFont.html#PIL.ImageFont.FreeTypeFont.getbbox) should be used instead: https://pillow.readthedocs.io/en/stable/releasenotes/9.2.0.html
> 
> As discussed in #8427, the kivy-ios and python-for-android Pillow recipes depend on versions of 8.* versions of Pillow, where `getbbox` isn't available. I believe this change will make deployment to mobile devices impossible.

`getbbox()` has been available since version 8.0.0: https://pillow.readthedocs.io/en/stable/reference/ImageFont.html#PIL.ImageFont.FreeTypeFont.getbbox

**Review by misl6 — COMMENTED** [review]
LGTM, but we will need a rebase after #8511 is merged.

**Comment by misl6** [comment]
Hi @DexerBR ,

Can you please rebase on top of `master` (As we now have a conflict)?

**Review by misl6 — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**
